tests/websocket.c

```diff
@@ -8,319 +8,537 @@
+                          avail_sz - nparsed > 1 ? 's' : ' ');
+                evbuffer_prepend(wt->rbuffer, tmp + nparsed, avail_sz - nparsed);
+            }
+            free(tmp);
+            if (wt->state == WS_SENT_HANDSHAKE &&  /* haven't encountered end of response yet */
+                wt->parser.upgrade && nparsed != (int)avail_sz) {
+                wt->debug("UPGRADE *and* we have some data left\n");
+                continue;
+            } else if (wt->state == WS_RECEIVED_HANDSHAKE) { /* we have the full response */
+                evbuffer_drain(wt->rbuffer, evbuffer_get_length(wt->rbuffer));
+            }
+            return;
+        }
+
+        case WS_SENT_FRAME: { /* waiting for frame response */
+            wt->debug("We're in WS_SENT_FRAME, just read a frame response. wt->rbuffer:\n");
+            evbuffer_debug_dump(wt, wt->rbuffer);
+            uint8_t flag_opcodes, payload_len;
+            if (evbuffer_get_length(wt->rbuffer) < 2) { /* not enough data */
+                wait_for_possible_read(wt);
+                return;
+            }
+            evbuffer_remove(wt->rbuffer, &flag_opcodes, 1); /* remove flags & opcode */
+            evbuffer_remove(wt->rbuffer, &payload_len, 1);  /* remove length */
+            evbuffer_drain(wt->rbuffer, (size_t)payload_len); /* remove payload itself */
+            process_message(wt, payload_len);
+
+            if (evbuffer_get_length(wt->rbuffer) == 0) { /* consumed everything */
+                if (wt->msg_received < wt->msg_target) { /* let's write again */
+                    wt->debug("our turn to write again\n");
+                    wt->state = WS_RECEIVED_HANDSHAKE;
+                    ws_enqueue_frame(wt);
+                } /* otherwise, we're done */
+                return;
+            } else {
+                wt->debug("there's still data to consume\n");
+                continue;
+            }
+            return;
+        }
+
+        default:
+            return;
+        }
+    }
+}
+
+
+static void
+wait_for_possible_read(struct worker_thread *wt) {
+    wt->debug("%s (wt=%p)\n", __func__, wt);
+    event_set(&wt->ev_r, wt->fd, EV_READ, websocket_can_read, wt);
+    event_base_set(wt->base, &wt->ev_r);
+    event_add(&wt->ev_r, NULL);
+}
+
+static void
+wait_for_possible_write(struct worker_thread *wt) {
+    wt->debug("%s (wt=%p)\n", __func__, wt);
+    event_set(&wt->ev_r, wt->fd, EV_WRITE, websocket_can_write, wt);
+    event_base_set(wt->base, &wt->ev_r);
+    event_add(&wt->ev_r, NULL);
+}
+
+static int
+ws_on_headers_complete(http_parser *p) {
+    struct worker_thread *wt = p->data;
+
+    wt->debug("%s (wt=%p)\n", __func__, wt);
+    // TODO
+    return 0;
+}
+
+static void
+ws_enqueue_frame_for_command(struct worker_thread *wt, char *cmd, size_t sz) {
+    unsigned char mask[4];
+    for (int i = 0; i < 4; i++) {
+        mask[i] = rand() & 0xff;
+    }
```

Comment:
would it make sense to generate the mask just once per client? we'd still be exercising this code but without calling `rand()` all the time.

Also: it would be great to have it as a command line option, maybe 0 → always disabled, 1 → always enabled, 2 → enabled half the time?

Reply:
The mask [should not be reused](https://datatracker.ietf.org/doc/html/rfc6455#page-33):
>  The masking key is a 32-bit value chosen at random by the client.
>  When preparing a masked frame, the client MUST pick a fresh masking
>   key from the set of allowed 32-bit values.